### DISCUSSION TITLE
.dll vs .cs scripts

### DISCUSSION THREAD

**Discussion by Michael21107** [scripthookvdotnet/scripthookvdotnet#1471] [Q&A] [answered]
if one makes the source code of his script public (eg. public repo), is there any benefit of making it a .dll file?

**Comment by kagikn** [comment]
yes, there are. If some script authors don't provide built binaries of their scripts, you'll need a program to build them from their solutions before you can use them in SHVDN. Note that SHVDN's compiler feature doesn't allow script authors to configure what assemblies their scripts should have references to, so you can't specify libraries such as LemonUI to add references when you put your scripts as .cs files. Script devs might want to configure more than what SHVDN's compiler provides, such as attribute injection of like, `modopt`.

**Reply by Michael21107** [reply]
so if one provides a .cs file as ''source-code'' (in a different folder that's not to be installed) along with regular to-be-installed .dll file, the .cs file won't work?

**Reply by kagikn** [reply, marked as answer]
Well, when there are multiple assemblies that has the same name and version (not the API version!), SHVDN skips some of them but how SHVDN skips such ones is not defined as a public documentation. Script format doesn't matter, can happen with any of `.cs`, `.vb`, and `.dll` files. https://github.com/scripthookvdotnet/scripthookvdotnet/blob/fd3eca5cc035686f8d513c69765769223764bfd9/source/core/ScriptDomain.cs#L506-L510

Some of the differences between placing .cs or .vb files and placing .dll files are like this;
- Placing .cs or .vb files
  - Less options when compiling scripts. Note that SHVDN compiles .cs or .vb files on the fly every time it gets loaded or reloaded.
  - There is no way to couple source files as one scripts
- Placing .dll files
  - Can use arbitrary libraries
  - Can use more language features when compiling

Script devs wouldn't be able to have their scripts achieve what they want to as a single .cs or .vb file when they want to do something advanced, but providing scripts as built .dll files can have more options to do.

**Reply by Michael21107** [reply]
gotcha, thanks a lot
